#1702 - Allow institution to specify custom regulatory body

sh16011993 opened the pull request:
- Updated the `institutionprofilecreation.json` form to add **Other** as a value to the **Institution regulating body dropdown** and the **Other Regulating Body** text field.

![image](https://user-images.githubusercontent.com/7859295/221971976-ab5e604b-8d72-44aa-a34c-1ce349c6ebe0.png)

- Updated the Create Institution API to persist the changes to the database.
- Added the `other_regulating_body` column to the database and created the appropriate database migration scripts along with the required SQL script to add the `other_regulating_body` column and also the SQL script to remove the column if the migration needs to be reverted.

![image](https://user-images.githubusercontent.com/7859295/221974946-0c169d0a-a15d-4c70-8f3c-a7d87cef9590.png)

- Updated the `institutionprofile.json` form to display **Other regulating body** field in the Institution profile on the AEST portal.

![image](https://user-images.githubusercontent.com/7859295/221978957-910af6f5-47fd-4477-8216-0677f811248e.png)

- Updated the Modify Institution API to persist the changes if the user updates the institution via the **Manage Institution** section on the UI.

**Updating the field:**
![image](https://user-images.githubusercontent.com/7859295/221980352-44fb6baf-3518-4e24-834c-b8888678fd86.png)

**After the update:**
![image](https://user-images.githubusercontent.com/7859295/221980628-ea5685bb-0842-4ebe-8f20-b8ffda77182f.png)

ann-aot commented on sources/packages/web/src/views/aest/institution/Profile.vue:
```diff
@@ -57,6 +57,11 @@
               :propertyValue="institutionProfileDetail.regulatingBody"
             />
+            <title-value
+              v-if="institutionProfileDetail.regulatingBody === `other`"
```
`v-if="institutionProfileDetail.regulatingBody === 'other'"`

sh16011993 replied:
If I will not use the single quotes, other won't be recognized in the template as a string with which `institutionProfileDetail.regulatingBody` is compared.

ann-aot replied:
no... my comment was to use single quotes instead of ```

sh16011993 replied:
My bad. I will replace the backticks with the single quote.

ann-aot commented on sources/packages/forms/src/form-definitions/institutionprofile.json:
```diff
@@ -3357,7 +3431,7 @@
     },
     {
-      "html": "<p><span style=\"background-color:rgb(255,255,255);color:rgba(0,0,0,0.87);\">By continuing into the StudentAidBC Application you agree to StudentAid BC’s Privacy policy.</span></p>",
+      "html": "<p><span style=\"background-color:rgb(255,255,255);color:rgba(0,0,0,0.87);\">By continuing into the StudentAidBC Application you agree to StudentAid BC&rsquo;s Privacy policy.</span></p>\n",
```
what is `&rsquo;`

sh16011993 replied:
BC&rsquo;s  =>  BC's

ann-aot replied:
did you add that? we can directly use `BC's` instead of HTML Alphanumeric value

ann-aot reviewed (commented):
please add description and UI screenshots

ann-aot commented on sources/packages/backend/apps/db-migrations/src/sql/Institution/Add-col-other-regulating-body.sql:
```diff
@@ -0,0 +1,4 @@
+-- Add other_regulating_body column to institutions table
+ALTER TABLE sims.institutions ADD COLUMN IF NOT EXISTS other_regulating_body VARCHAR(100);
```
looks like you don't have the SQL formater

ann-aot replied:
same for all .sql files

dheepak-aot replied:
Yes @ann-aot is right. It looks like this if formatted with the sql formatter. 
![image](https://user-images.githubusercontent.com/54600590/221716216-a3c30a8a-d372-440e-98e5-53e5ffe7c302.png)

sh16011993 replied:
Thank You Ann and Deepak. Fixed it.

ann-aot commented on sources/packages/backend/apps/api/src/route-controllers/institution/models/institution.dto.ts:
```diff
@@ -114,6 +120,7 @@ export class InstitutionProfileAPIOutDTO extends InstitutionContactAPIOutDTO {
   primaryEmail: string;
   website: string;
   regulatingBody: string;
+  otherRegulatingBody: string;
```
`otherRegulatingBody?: string;`

sh16011993 replied:
Sounds good. Will do the change.

ann-aot commented on sources/packages/backend/apps/api/src/services/institution/institution.service.model.ts:
```diff
@@ -34,6 +35,7 @@ export interface InstitutionFormModel {
   primaryEmail: string;
   website: string;
   regulatingBody: string;
+  otherRegulatingBody: string;
```
`otherRegulatingBody?: string;` as its optional

sh16011993 replied:
Sounds good. Will do the change.

ann-aot commented on sources/packages/forms/src/form-definitions/institutionprofile.json:
```diff
@@ -1,12 +1,10 @@
-    "common"
-  ],
+  "type": "form",
+  "display": "form",
```
please enable the formatter  prettier and eslint

ann-aot replied:
same for all form form definition

dheepak-aot replied:
Also please use the JSON formatter in vs code as default for all the *.json files. 
![image](https://user-images.githubusercontent.com/54600590/221717336-11ef84d7-57c0-4218-b120-9685785af1d5.png)

ann-aot commented on sources/packages/web/src/services/http/dto/Institution.dto.ts:
```diff
@@ -27,6 +27,7 @@ export interface InstitutionProfileAPIOutDTO
   primaryEmail: string;
   website: string;
   regulatingBody: string;
+  otherRegulatingBody: string;
```
`otherRegulatingBody?: string;` as its optional

ann-aot commented on sources/packages/web/src/services/http/dto/Institution.dto.ts:
```diff
@@ -37,6 +38,7 @@ export interface InstitutionProfileAPIInDTO extends InstitutionContactAPIInDTO {
   primaryEmail: string;
   website: string;
   regulatingBody: string;
+  otherRegulatingBody: string;
```
`otherRegulatingBody?: string;` as its optional

sh16011993 replied:
Sounds good. Will do the change.

ann-aot reviewed (commented):
Good start @sh16011993 👍 Added some comments

dheepak-aot commented on sources/packages/backend/apps/api/src/route-controllers/institution/models/institution.dto.ts:
```diff
@@ -37,6 +37,9 @@ export class CreateInstitutionAPIInDTO {
   regulatingBody: string;
+  @ValidateIf((e) => e.regulatingBody === "other")
+  @IsNotEmpty()
+  otherRegulatingBody: string;
```
Not related to your PR. @andrewsignori-aot @andrepestana-aot @ann-aot we may have to eventually apply more specific validation (like @MaxLength) to the string to avoid API being by a huge string data.

andrewsignori-aot replied:
This is related to the PR @dheepak-aot.
We do not have a need to leave it for later anymore. For every new filed we need to be as specific as possible.
For this one, we can limit to the same length available for DB.

sh16011993 replied:
Sounds Good. Thanks Andrew, I will add it.

sh16011993 commented on sources/packages/web/src/services/http/dto/Institution.dto.ts:
```diff
@@ -98,6 +100,7 @@ export interface CreateInstitutionAPIInDTO {
   primaryEmail: string;
   website: string;
   regulatingBody: string;
+  otherRegulatingBody: string;
```
Sounds good. Will do the change.

sh16011993 commented on sources/packages/backend/libs/sims-db/src/entities/institution.model.ts:
```diff
@@ -63,11 +63,17 @@ export class Institution extends RecordDataModel {
+  @Column({
+    name: "other_regulating_body",
+  })
+  otherRegulatingBody: string;
```
Sounds good. Will do the change.

dheepak-aot commented on sources/packages/backend/apps/db-migrations/src/sql/Institution/Add-col-other-regulating-body.sql:
```diff
@@ -0,0 +1,4 @@
+-- Add other_regulating_body column to institutions table
+ALTER TABLE sims.institutions ADD COLUMN IF NOT EXISTS other_regulating_body VARCHAR(100);
+
+COMMENT ON COLUMN sims.institutions.other_regulating_body IS 'Allows the user to enter the description if the option other is selected from the regulating body dropdown';
```
The comment can avoid client specific content like `Allows the user to enter the description` and just convey the business context here. (e.g. Other institution regulating body)

sh16011993 replied:
Thanks Dheepak. Will make the change.

dheepak-aot commented:
Great start @sh16011993 . Please have a look at the comments. 

On PR please make sure add at least a quick summary of the efforts and if there are UI changes in the PR, please add screenshots.

Here are some examples.
https://github.com/bcgov/SIMS/pull/1747
https://github.com/bcgov/SIMS/pull/1754

andrewsignori-aot commented on sources/packages/backend/apps/db-migrations/src/sql/Institution/Add-col-other-regulating-body.sql:
```diff
@@ -0,0 +1,4 @@
+-- Add other_regulating_body column to institutions table
+ALTER TABLE sims.institutions ADD COLUMN IF NOT EXISTS other_regulating_body VARCHAR(100);
+
+COMMENT ON COLUMN sims.institutions.other_regulating_body IS 'Allows the user to enter the description if the option other is selected from the regulating body dropdown';
```
Please add a period at the end of the sentence.

sh16011993 replied:
Thanks. Will do.

andrewsignori-aot reviewed (commented):
Nice and great first PR. @ann-aot and @dheepak-aot covered pretty much everything. I will be waiting for the changes to review again.

andrepestana-aot reviewed (commented):
Great job @sh16011993 . The other devs already addressed the issues. One thing I noticed: please let the person who comment to resolve the conversation.

sh16011993 commented on sources/packages/backend/libs/sims-db/src/entities/institution.model.ts:
```diff
@@ -63,11 +63,17 @@ export class Institution extends RecordDataModel {
+  })
+  otherRegulatingBody: string;
+
   @Column({
```
Sounds good. Will do the change

sh16011993 commented:
> Great job @sh16011993 . The other devs already addressed the issues. One thing I noticed: please let the person who commented to resolve the conversation.

Actually I was talking to @ann-aot at the time I resolved the specific conversation with her.

sh16011993 commented:
> Nice and great first PR. @ann-aot and @dheepak-aot covered pretty much everything. I will be waiting for the changes to review again.

@andrewsignori-aot I have updated the code with the review comments.

dheepak-aot commented on sources/packages/backend/apps/api/src/route-controllers/institution/models/institution.dto.ts:
```diff
@@ -102,6 +108,9 @@ export class InstitutionProfileAPIInDTO extends InstitutionContactAPIInDTO {
   @IsNotEmpty()
   regulatingBody: string;
+  @ValidateIf((e) => e.regulatingBody === "other")
+  @IsNotEmpty()
```
Same max length validation needs to added here as well.

sh16011993 replied:
I apologize, I missed it.

dheepak-aot replied:
No problem.

dheepak-aot commented on sources/packages/backend/libs/sims-db/src/entities/institution.model.ts:
```diff
@@ -63,11 +63,20 @@ export class Institution extends RecordDataModel {
   regulatingBody: string;
 
+  /**
+   * Other regulating body
```
Period missing in the comment. Also please try to use the same comment used for the column.

![image](https://user-images.githubusercontent.com/54600590/222004569-3bb113b8-839d-465d-837d-b5f65fa63c31.png)

dheepak-aot commented:
Thanks for doing the changes. Added 2 more comments.

dheepak-aot reviewed (approved):
Thanks for doing the changes @sh16011993. Nice beginning.

andrepestana-aot reviewed (approved):
LGTM

andrewsignori-aot commented on sources/packages/forms/src/form-definitions/institutionprofilecreation.json:
```diff
@@ -1025,140 +1025,217 @@
+                  "tableView": true,
+                  "inputType": "text",
+                  "inputMask": "",
+                  "label": "Other Regulating Body",
```
Looking at the other fields on the form it looks like the case should be "Other regulating body".

andrewsignori-aot commented on sources/packages/forms/src/form-definitions/institutionprofile.json:
```diff
@@ -562,140 +562,216 @@
+                  "tableView": true,
+                  "inputType": "text",
+                  "inputMask": "",
+                  "label": "Other Institution Regulating Body",
```
Looking at the other fields on the form it looks like the case should be "Other institution regulating body".

andrewsignori-aot commented on sources/packages/backend/apps/api/src/route-controllers/institution/models/institution.dto.ts:
```diff
@@ -37,6 +39,10 @@ export class CreateInstitutionAPIInDTO {
   regulatingBody: string;
+  @ValidateIf((e) => e.regulatingBody === "other")
+  @IsNotEmpty()
+  @MaxLength(OTHER_REGULATING_BODY_MAX_LENGTH)
```
Since we created the below file I believe that we can keep adding these constants there, unless they are beyond the API scope.
```
sources\packages\backend\apps\api\src\constants\class-validator-constants.ts
```
@andrepestana-aot @ann-aot @dheepak-aot FYI

andrewsignori-aot reviewed (commented):
Thanks for doing the changes. Please take a look on the last minor comments.

andrewsignori-aot reviewed (approved):
Thanks for doing the changes, it looks good 👍

ann-aot reviewed (approved):
Good start and good work @sh16011993 👍  Thanks for doing the changes
